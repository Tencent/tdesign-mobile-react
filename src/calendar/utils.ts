import { TCalendarValue } from './type';

export function getMonthByOffset(date: TCalendarValue, offset: number) {
  const innerDate = new Date(date);
  innerDate.setMonth(innerDate.getMonth() + offset);
  innerDate.setDate(1);
  return innerDate;
}

export function getYearByOffset(date: TCalendarValue, offset: number) {
  const innerDate = new Date(date);
  innerDate.setFullYear(innerDate.getFullYear() + offset);
  innerDate.setDate(1);
  return innerDate;
}

export const getPrevMonth = (date: TCalendarValue) => getMonthByOffset(date, -1);
export const getNextMonth = (date: TCalendarValue) => getMonthByOffset(date, 1);
export const getPrevYear = (date: TCalendarValue) => getYearByOffset(date, -1);
export const getNextYear = (date: TCalendarValue) => getYearByOffset(date, 1);
