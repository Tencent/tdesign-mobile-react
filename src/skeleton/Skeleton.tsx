import React, { useState, useEffect } from 'react';
import classNames from 'classnames';
import { isArray, isNumber } from 'lodash-es';
import { SkeletonRowCol, SkeletonRowColObj, TdSkeletonProps } from './type';
import { StyledProps, Styles } from '../common';
import { skeletonDefaultProps } from './defaultProps';
import { pxCompat } from '../_util/helper';
import parseTNode from '../_util/parseTNode';
import useDefaultProps from '../hooks/useDefaultProps';
import { usePrefixClass } from '../hooks/useClass';

export interface SkeletonProps extends TdSkeletonProps, StyledProps {}

const ThemeMap: Record<TdSkeletonProps['theme'], SkeletonRowCol> = {
  avatar: [{ type: 'circle', size: '48px' }],
  image: [{ type: 'rect', size: '72px' }],
  text: [
    [
      { width: '24%', height: '16px', marginRight: '16px' },
      { width: '76%', height: '16px' },
    ],
    1,
  ],
  paragraph: [1, 1, 1, { width: '55%' }],
};

const Skeleton: React.FC<SkeletonProps> = (props) => {
  const skeletonClass = usePrefixClass('skeleton');
  const { className, style, children, animation, delay, loading, rowCol, theme } = useDefaultProps<SkeletonProps>(
    props,
    skeletonDefaultProps,
  );

  const renderCols = (_cols: Number | SkeletonRowColObj | Array<SkeletonRowColObj>) => {
    let cols: Array<SkeletonRowColObj> = [];
    if (isArray(_cols)) {
      cols = _cols;
    } else if (isNumber(_cols)) {
      cols = new Array(_cols).fill({ type: 'text' });
    } else {
      cols = [_cols as SkeletonRowColObj];
    }

    const getColItemClass = (obj: SkeletonRowColObj) =>
      classNames(`${skeletonClass}__col`, `${skeletonClass}--type-${obj.type || 'text'}`, {
        [`${skeletonClass}--animation-${animation}`]: animation,
      });

    const getColItemStyle = (obj: SkeletonRowColObj): Styles => {
      const styleName = [
        'width',
        'height',
        'marginRight',
        'marginLeft',
        'margin',
        'size',
        'background',
        'backgroundColor',
        'borderRadius',
      ];
      const style: Styles = {};
      styleName.forEach((name) => {
        if (name in obj) {
          const px = pxCompat(obj[name]);
          if (name === 'size') {
            [style.width, style.height] = [px, px];
          } else {
            style[name] = px;
          }
        }
      });
      return style;
    };

    return cols.map((obj, index) => (
      <div key={index} className={getColItemClass(obj)} style={getColItemStyle(obj)}>
        {parseTNode(obj.content)}
      </div>
    ));
  };

  const renderRowCol = (_rowCol?: SkeletonRowCol) => {
    const renderedRowCol: SkeletonRowCol = _rowCol || rowCol;

    return renderedRowCol.map<React.ReactNode>((item, index) => (
      <div key={index} className={`${skeletonClass}__row`}>
        {renderCols(item)}
      </div>
    ));
  };

  const [isLoading, setIsLoading] = useState(loading);

  useEffect(() => {
    if (delay > 0 && !loading) {
      const timeout = setTimeout(() => {
        setIsLoading(loading);
      }, delay);
      return () => clearTimeout(timeout);
    }

    setIsLoading(loading);
  }, [delay, loading]);

  if (!isLoading) {
    return <>{parseTNode(children)}</>;
  }

  const childrenContent: React.ReactNode[] = [];

  // 保持优先级： rowCol > theme，增加默认值兜底
  if (rowCol) {
    childrenContent.push(renderRowCol(rowCol));
  } else if (props.theme) {
    childrenContent.push(renderRowCol(ThemeMap[theme]));
  } else if (!theme && !rowCol) {
    // 什么都不传时，传入默认 rowCol
    childrenContent.push(
      renderRowCol([
        [
          { width: '24%', height: '16px', marginRight: '16px' },
          { width: '76%', height: '16px' },
        ],
        1,
      ]),
    );
  }

  return (
    <div className={className} style={style}>
      {childrenContent}
    </div>
  );
};

Skeleton.displayName = 'Skeleton';

export default Skeleton;
