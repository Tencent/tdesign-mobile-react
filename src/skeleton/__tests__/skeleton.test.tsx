import { render } from '@test/utils';
import { act } from '@testing-library/react';
import { expect, it, describe, vi } from 'vitest';
import React from 'react';
import Skeleton, { SkeletonProps } from '../index';

describe('Skeleton', () => {
  describe('props', () => {
    it(':theme', async () => {
      const themes = ['avatar', 'image', 'text', 'paragraph'] as const;
      const themeRowAndCol = {
        avatar: [1, 1],
        image: [1, 1],
        text: [2, 3],
        paragraph: [4, 4],
      };

      const checkThemes = async (theme: SkeletonProps['theme']) => {
        const { container } = render(<Skeleton theme={theme} />);

        const rows = container.querySelectorAll('.t-skeleton__row');
        const cols = container.querySelectorAll('.t-skeleton__col');
        expect(rows.length).toBe(themeRowAndCol[theme][0]);
        expect(cols.length).toBe(themeRowAndCol[theme][1]);
      };

      themes.forEach(checkThemes);
    });

    it(':rowCol', async () => {
      const customRowCols = {
        first: {
          rowCols: [{ size: '48px', type: 'circle' }],
          row: 1,
          col: 1,
        },

        second: {
          rowCols: [{ size: '48px', type: 'rect' }],
          row: 1,
          col: 1,
        },
        third: {
          rowCols: [{ width: '50%' }, { width: '100%' }],
          row: 2,
          col: 2,
        },
        fourth: {
          rowCols: [
            [
              { width: '48px', height: '48px', borderRadius: '6px' },
              { width: '48px', height: '48px', borderRadius: '6px' },
              { width: '48px', height: '48px', borderRadius: '6px' },
              { width: '48px', height: '48px', borderRadius: '6px' },
              { width: '48px', height: '48px', borderRadius: '6px' },
            ],
            [
              { width: '48px', height: '16px', borderRadius: '3px' },
              { width: '48px', height: '16px', borderRadius: '3px' },
              { width: '48px', height: '16px', borderRadius: '3px' },
              { width: '48px', height: '16px', borderRadius: '3px' },
              { width: '48px', height: '16px', borderRadius: '3px' },
            ],
          ],
          row: 2,
          col: 10,
        },
      };

      const checkRowCols = async (key: string) => {
        const { rowCols, row, col } = customRowCols[key];
        const { container } = render(<Skeleton rowCol={rowCols} />);

        const rows = container.querySelectorAll('.t-skeleton__row');
        const cols = container.querySelectorAll('.t-skeleton__col');
        expect(rows.length).toBe(row);
        expect(cols.length).toBe(col);
      };
      Object.keys(customRowCols).forEach(checkRowCols);
    });

    it(':animation', async () => {
      const animations = [
        {
          title: '渐变加载效果',
          value: 'gradient' as SkeletonProps['animation'],
          loading: true,
        },
        {
          title: '闪烁加载效果',
          value: 'flashed' as SkeletonProps['animation'],
          loading: true,
        },
      ];

      const checkAnimation = async (animation) => {
        const { container } = render(
          <Skeleton theme="paragraph" animation={animation.value} loading={animation.loading} />,
        );
        expect(container.querySelector('.t-skeleton__col')).toHaveClass(`t-skeleton--animation-${animation.value}`);
      };

      animations.forEach(checkAnimation);
    });

    it(':delay', async () => {
      vi.useFakeTimers();

      const { container, rerender, getByTestId } = render(
        <Skeleton loading={false} delay={1000}>
          <div data-testid="skeleton-content">
            <p>设置最短延迟响应时间，低于响应时间的操作不显示加载状态。</p>
          </div>
        </Skeleton>,
      );

      // loading = false，不应显示骨架屏，子内容应显示
      const rows1 = container.querySelectorAll('.t-skeleton__row');
      expect(rows1.length).toBe(0);
      expect(getByTestId('skeleton-content')).toBeTruthy();

      // 快进时间 1000ms（达到 delay 时间）
      vi.advanceTimersByTime(1000);

      // loading = false，即使到达 delay 时间，也不应显示骨架屏
      const rows2 = container.querySelectorAll('.t-skeleton__row');
      expect(rows2.length).toBe(0);
      expect(getByTestId('skeleton-content')).toBeTruthy();

      // 将 loading 设置为 true
      rerender(
        <Skeleton loading={true} delay={1000}>
          <div data-testid="skeleton-content">
            <p>设置最短延迟响应时间，低于响应时间的操作不显示加载状态。</p>
          </div>
        </Skeleton>,
      );

      // 快进 500ms，但是 未达 delay 时间，不应显示骨架屏
      await act(async () => {
        vi.advanceTimersByTime(500);
      });

      const rows3 = container.querySelectorAll('.t-skeleton__row');
      expect(rows3.length).toBe(0);
      expect(getByTestId('skeleton-content')).toBeTruthy();

      // 继续时间 500ms，达到 delay 时间，应显示骨架屏
      await act(async () => {
        vi.advanceTimersByTime(500);
      });
      const rows4 = container.querySelectorAll('.t-skeleton__row');
      expect(rows4.length).toBe(2);
      expect(() => getByTestId('skeleton-content')).toThrow(); // getByTestId 在元素不存在时会直接跑出错误

      vi.useRealTimers();
    });

    it(': other sense', () => {
      // 无主题无rowCol时，使用默认值兜底
      const { container } = render(<Skeleton theme={null} rowCol={null} />);

      const rows = container.querySelectorAll('.t-skeleton__row');
      const cols = container.querySelectorAll('.t-skeleton__col');
      expect(rows.length).toBe(2);
      expect(cols.length).toBe(3);

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
