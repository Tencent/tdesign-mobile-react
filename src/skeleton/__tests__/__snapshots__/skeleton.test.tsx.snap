// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Skeleton > :props other > no theme and no rowCol 1`] = `
<div>
  <div
    class="t-skeleton__row"
  >
    <div
      class="t-skeleton__col t-skeleton--type-text t-skeleton--animation-none"
      style="width: 24%; height: 16px; margin-right: 16px;"
    />
    <div
      class="t-skeleton__col t-skeleton--type-text t-skeleton--animation-none"
      style="width: 76%; height: 16px;"
    />
  </div>
  <div
    class="t-skeleton__row"
  >
    <div
      class="t-skeleton__col t-skeleton--type-text t-skeleton--animation-none"
    />
  </div>
</div>
`;
